### ISSUE TITLE
different server-side rendering behaviour between v2 and v3

### ISSUE THREAD

**Issue by cusxio** [apollographql/react-apollo#3338]
## Current Behaviour

`v2` and `v3` of `react-apollo` have vastly different behaviour/timings server side.

On `v3`, it seems that `react-apollo` has an additional render pass/intermediate state on the server side that breaks `v2` behaviour.

The **additional render pass** seems to output the following:

https://github.com/apollographql/react-apollo/blob/baf9a7648ac4ef2fc486f27719736b0d4e198340/packages/hooks/src/data/QueryData.ts#L172-L177

## Reproduction:

`v2`: https://codesandbox.io/s/with-apollo-xrj2c

`v3`: https://codesandbox.io/s/with-apollo-8luew

Check out the Terminal output difference of both `v2` and `v3`, 

`v2` only has two render pass, but `v3` has three render pass.

The additional render pass in `v3` currently breaks our app, as we rely on the `variable` data being there.

## Expected Behaviour

`v3` server side rendering behaviour should be similar to `v2`

**Comment by NinjaOnRails** [comment]
I'm having a similar issue [here]( https://github.com/zeit/next.js/pull/8180#issuecomment-520226158)

**Comment by HugoGiraudel** [comment]
I can confirm. `react-apollo@^3.0.0` has a rendering pass on the server with `loading` being `true`.

**Comment by youfoundron** [comment]
Also seeing this issue. Updating Next did not fix this for me.

**Comment by flipxfx** [comment]
We are seeing the same issue of the data on SSR still showing `loading: true`

**Comment by mikebm** [comment]
Also experiencing this. The SSR is only returning our loading components.

**Comment by mikebm** [comment]
I've been debugging this a little bit today and it appears to be isolated to `network-only` fetchPolicy. Still unsure why and will keep digging, but one could potentially use `cache-and-network` as workaround. I personally went with `network-only` because I was worried about the client side cache getting too big for my particular case since there is currently no TTL support.

**Comment by mikebm** [comment]
@hwillson I was able to "resolve" my issue where a fetchPolicy of `network-only` causes only the loading component to render during SSR. It appears to be due to the initial observable getting initialized in execute() before fetchData is called. I was able to adjust the `updateObservableQuery` function with the below code:
```
	if (this.context.renderPromises && (newObservableQueryOptions.fetchPolicy === 'network-only' || newObservableQueryOptions.fetchPolicy === 'cache-and-network')) {
		newObservableQueryOptions.fetchPolicy = 'cache-first';
	}
```

This adjusts the options to be consistent with what fetchData does. Ideally, i would think this logic would live in `initializeObservableQuery`, but the code in `updateObservableQuery` would immediately blow away the options.  

I'd also like to point out that `getOptions` is pretty inefficient as it always returns a new object and is called all over the place.

I am not sure this addresses the original posters complaint of an additional "loading" render during SSR, but how it works now makes it more consistent with how it works on the client. The main issue I can see is that `network-only` fetches with useQuery results in no data returned and instead whatever component you have rendering in your loading event which some form of my above findings would resolve.

**Comment by hwillson** [comment]
Hi all - question regarding the OP; would people prefer to see the initial `loading` state of `true` when using SSR, or not? React Apollo 2 skipped displaying it (thereby skipping the additional render), but we have had requests to enable it (hence the reason why it's in RA 3). I see now that this wasn't mentioned in the breaking changes (argh 😳), so I can revert things back to the way they were with RA 2; that is, no initial `loading: true` when using SSR. Having the additional `loading: true` step does add some extra flexibility though, so I'd love to hear what everyone prefers before (possibly) changing this back.

Thanks for looking into this @mikebm! I don't think your change will resolve the initially reported issue, but your suggested change makes a lot of sense regardless. Any chance you'd be interested in submitting that as a PR?

**Comment by mikebm** [comment]
I personally think the loading state should stay. It makes things consistent.  In my PR, which I just opened, also addresses an issue where disabling SSR on a query was still resulting in the query being fired but the result ignored. In this case, I had to still return a state when the hook was executed, so I chose to use the ssrLoading event.

**Comment by cusxio** [comment]
> Having the additional loading: true step does add some extra flexibility though, so I'd love to hear what everyone prefers before (possibly) changing this back.

Is there an example of a use case where `loading: true` is desired?

---

Currently in our app, because `loading: true` during the initial render, and we fail to account for it, our app does a redirection because it can't find the needed data (i.e. `currentUser`).

---

**Question**:

Correct me if I'm wrong, but is this how server-side rendering  work in Apollo?:

1. Initial render, recursively finds all the `Query` components in the app via calling `renderToString`.
2. Wait for all promises to resolve to get the data.
3. Call `renderToString` again with the data to get the correct markup.

**Comment by HugoGiraudel** [comment]
Hello there. 👋 

We have been using Apollo for over 2 years at N26 and have been delighted with it so far. We were stoked to update to v3 and migrate to hooks for more elegant components. However… this breaking change is quite a bummer in my opinion. :(

I might be missing something, but I genuinely don’t understand the benefit of shipping a loading state to the client. If the server renders with `loading: true`, that means it will generate markup for the loading state of all querying components. Correct?

The point of SSR — at least the way we use SSR — is that the page should be readable/usable when hitting the client, so much that we could skip shipping the bundles entirely and have everything there.  This change makes that impossible as displaying loaders is not quite achieving that.

It would be lovely if we could restore the v2 behaviour somehow. 😊 ✨

**Comment by flipxfx** [comment]
I completely agree with @HugoGiraudel and @cusxio. I think the previous default makes more sense from an SSR "ready for client" perspective. We rely on data being ready and not loading, like `currentUser`. I don't see the point of `loading = true` if we are using `getDataFromTree` in SSR before returning the SSR pages to the client. 

I'm not sure what the use case is for `loading = true`, but it seems like it at least shouldn't be the default. If there is a use for it, maybe have it as a feature flag in Apollo init.

❤️

**Comment by mikebm** [comment]
I think the real question is why are your components stuck in a loading state? Ideally your component would never render to the client the loading component during a SSR, except maybe when SSR is off for the query. I feel like you all might be running into a different bug that prevented the query from finishing and either giving an error in data. What fetchPolicy are you guy's using?

**Comment by OllieJennings** [comment]
@hwillson It seems like Cookies are not being sent when on SSR unless you manually include `req.headers` (this isn't required on client, just server)

**Comment by mikebm** [comment]
> @hwillson It seems like Cookies are not being sent when on SSR unless you manually include `req.headers` (this isn't required on client, just server)

That's normal. RESTDataSource, which uses fetch internally, isn't aware of express/hapi/what have you's request object, so it cannot pull cookies off of it to forward. You have to do it manually. On each request, when you build your Apollo Client on the server, you need to pass headers from your request object to createHttpLink.

**Comment by OllieJennings** [comment]
@mikebm got ya, cheers for the explanation

**Comment by flipxfx** [comment]
The components aren’t stuck in loading state, they just begin in loading state then switch to done after it gets the data from cache. But we would like SSR to skip the loading state completely as we already have all the data loaded. It happens with the default fetch policy. 

To confirm we aren’t having the cookies issue as we already manually attach this to the request.

**Comment by mikebm** [comment]
> The components aren’t stuck in loading state, they just begin in loading state then switch to done after it gets the data from cache. But we would like SSR to skip the loading state completely as we already have all the data loaded. It happens with the default fetch policy.
> 
> To confirm we aren’t having the cookies issue as we already manually attach this to the request.

So, you are saying the SSR is only returning the loading state to your component and never loading: false, with data populated? In this scenario, you are also saying that your apollo cache was populated with the data.  

I was definitely not experiencing this with the default cache policy, which I believe is cache-first. It would go through the loading state, but once the data was resolved, my component was re-rendered and loading set to false.

**Comment by flipxfx** [comment]
We are using getDataFromTree and passing a Apollo client with initial cached data to it.

**Comment by hwillson** [comment]
I'd really like to get this issue resolved before publishing 3.1.0, so let's all hash out a plan of attack together. 

## Background

In React Apollo 2, the `Query` component returns `null` for the first render triggered during SSR (when the query is loading on the server). So when a component like the following

```jsx
const SomeComponent = () => {
  return (
    <Query query={CAR_QUERY}>
      {({ loading, data }) => {
        if (!loading) {
          const { make, model, vin } = data.cars[0];
          return (
            <div>
              {make}, {model}, {vin}
            </div>
          );
        }
      }}
    </Query>
  )
};
```

is rendered on the server, its first pass essentially renders a component that looks like:

```jsx
const SomeComponent = () => null;
```

One advantage render prop functions have over hooks in this situation is that they encapsulate all query result handling, and can then decide if they should or should not return `null` to the parent component. In other words, if the returned query result is `null` because we're handling the first step of SSR, the following destructuring won't break:

```jsx
{({ loading, data }) => {
```

This is because internally before running the render prop function with the results of the initial SSR query call (which we haven't received yet from Apollo Client), we just return `null` instead of triggering the render prop function.

When it comes to the React Apollo 3 hooks approach however, `useQuery` calls generally look like this:

```jsx
const SomeComponent = () => {
  const { loading, data } = useQuery(CAR_QUERY);
  if (!loading) {
    const { make, model, vin } = data.cars[0];
    return (
      <div>
        {make}, {model}, {vin}
      </div>
    );
  }
  return null;
};
```

Consumers of `useQuery` are deciding how to handle destructuring, and this isn't something React Apollo can control (unlike the render prop function approach). During the first SSR pass, since we can't hold up a render, we have to return something. The render prop function approach of returning `null` doesn't work quite as nicely, since most people aren't expecting a `null` result from `useQuery`, and their destructuring will fail. To workaround this, React Apollo 3 currently returns a `loading: true` based result for the first SSR pass, which is accurately mirroring what's really happening behind the scenes. Unfortunately, as mentioned in several comments in this thread, it can lead to un-necessary `loading: true` states being shipped to the client.

## Solution

It is possible to add extra logic to block `loading: true` results from being sent to the client, but that extra work can be a pain (especially since this wasn't necessary in React Apollo 2).

Looking ahead for ways to change this, would people prefer that we drop the initial SSR `loading: true` result, and instead return `null` from `useQuery` as the query is loading? To simulate how the `Query` comonent works in React Apollo 2, you could then update your component like:

```jsx
const SomeComponent = () => {
  const result = useQuery(CAR_QUERY);
  if (result) {
    const { loading, data } = result;
    if (!loading) {
      const { make, model, vin } = data.cars[0];
      return (
        <div>
          {make}, {model}, {vin}
        </div>
      );
    }
  }
  return null;
};
```

If people don't care about SSR, they don't have to worry about checking for `null`, since non-SSR results can never be `null`. Would something like this be helpful?

**Comment by mikebm** [comment]
I'd still like to narrow down the "it can lead to un-necessary loading: true states being shipped to the client." as this implies the SSR didn't wait for the query to resolve and return data or an error which would have caused the `useQuery` to trigger a re-render with a more completed state.

**Comment by hwillson** [comment]
@mikebm When `useQuery` is fired on the server, it has to complete the initial render. As part of that initial render, it preps and fires the query, and gets ready to receive the result. While it's waiting for the result, it has to return something to complete the render. React Apollo 2 is returning `null`, React Apollo 3 is return a result object with `loading: true`. We can't hold up returning that initial result, and if people are shipping that initial result to the client, it gets out.

**Comment by mikebm** [comment]
> @mikebm When `useQuery` is fired on the server, it has to complete the initial render. As part of that initial render, it preps and fires the query, and gets ready to receive the result. While it's waiting for the result, it has to return something to complete the render. React Apollo 2 is returning `null`, React Apollo 3 is return a result object with `loading: true`. We can't hold up returning that initial result, and if people are shipping that initial result to the client, it gets out.

I understand this, but the server shouldn't be returning those initial results to the client because the SSR will wait for all the renderPromises to resolve, which once they do, it will re-trigger the components to re-render in a non-loading state.  So, my question is, aside from the bug I recently fixed, how is it that these components are being returned to the client in a loading state when using the proper SSR features in this library.

**Comment by hwillson** [comment]
@mikebm For sure, and great question. The original reproduction links in this issue thread show the `loading: true` differences between React Apollo 2 and 3, but they don't really show an actual issue. The content shipped to the client is still valid and its `loading` state is `false`. The differences between RA 2 and 3 can be explained (https://github.com/apollographql/react-apollo/issues/3338#issuecomment-528462945), but ultimately the `loading: true` state shouldn't be trickling out to the client.

So ... I guess we need more details on what the actual issue is there then. Is anyone here able to put together a reproduction that shows how the additional loading state is causing grief? Please note that if you want to use the next.js [`with-apollo`](https://github.com/zeit/next.js/tree/canary/examples/with-apollo) demo app as a starting point, they just released a new version yesterday that is using `@apollo/react-hooks` and `@apollo/react-ssr`.

**Comment by flipxfx** [comment]
It looks like the `with-apollo` example is actually returning `loading: false` correctly. The `with-apollo` example uses the HOC for page components, but we have the Apollo client built into `_app.js`.

I noticed in the `with-apollo` example, it has `console.warn('This withApollo HOC only works with PageComponents.')`

So is it not possible to setup Apollo client in `_app.js`? Or maybe we just have it configure incorrectly. We currently have it built into `_app.js` because we want Apollo client, authentication, and `currentUser` automatically on every page. If it is possible, I think it'd be good to have an example of setting up Apollo client in `_app.js`

**Comment by seantimm** [comment]
@hwillson - I need to work on getting together a concrete example, but I'm seeing the issue with SSR'd components in the loading state in composed queries:

```gql
query GET_HEADER_FEATURED {
  header1: posts(categories: ["header-featured"], everything: true, size: 1) {
    ...postFields
  }

  header2: posts(categories: ["header-featured-2"], everything: true, size: 2) {
    ...postFields
  }

  header3: posts(categories: ["header-featured-3"], everything: true, size: 1) {
    ...postFields
  }
}
```

If I comment out header2 and header3, it does the SSR correctly.

**Comment by kedarguy** [comment]
> @mikebm For sure, and great question. The original reproduction links in this issue thread show the `loading: true` differences between React Apollo 2 and 3, but they don't really show an actual issue. The content shipped to the client is still valid and its `loading` state is `false`. The differences between RA 2 and 3 can be explained ([#3338 (comment)](https://github.com/apollographql/react-apollo/issues/3338#issuecomment-528462945)), but ultimately the `loading: true` state shouldn't be trickling out to the client.
> 
> So ... I guess we need more details on what the actual issue is there then. Is anyone here able to put together a reproduction that shows how the additional loading state is causing grief? Please note that if you want to use the next.js [`with-apollo`](https://github.com/zeit/next.js/tree/canary/examples/with-apollo) demo app as a starting point, they just released a new version yesterday that is using `@apollo/react-hooks` and `@apollo/react-ssr`.

I created a repo [here](https://codesandbox.io/s/with-apollo-70b0k) for the related issue I opened #3390 that shows the error we get in getDataFromTree because of this issue. in this simple case it still goes through somehow, In our app it breaks the SSR.

I will try to create a simple repo where it actually breaks the SSR.

**Comment by lifeiscontent** [comment]
this might be related: https://github.com/apollographql/react-apollo/issues/3486

also @flipxfx you can use apollo with _app.js the reason why the new example is using page components only is due to how the new Next.js 9 automatic prerendering works. If you wrap `_app.js` you automatically opt-out of it on every page.

**Comment by lifeiscontent** [comment]
ping @hwillson please take a look at https://github.com/apollographql/react-apollo/issues/3338#issuecomment-530515129 seems like there are some issues related / around it.

**Comment by lasseborly** [comment]
> I found that it happens when we query for client data with @client directive, and then in a nested component we query for remote data.

I can reproduce the same behavior @kedarguy.

**Comment by carlows** [comment]
Btw this is also happening to me when using the `@client` directive but from v2, I don't think this is related to v3 changes specifically

**Comment by mariohoyos92** [comment]
I agree with the sentiment that when I use SSR I am expecting the page to come with all queries resolved already. I'm sure a ton of us use SSR for SEO purposes and it doesn't do us much good in that regard to show an empty page while we fetch data. The advantage to SSR is handling this on the server and delivering a fleshed out page, otherwise might as well just run with create-react-app and not worry about SSR.

**Comment by kaanozcan** [comment]
If renderToStringWithData would output each of the renders one by one we could do pre-rendering. Contract would be changed alot but we could send responses as fast as possible to the client and the people who doesn't want to do pre-rendering would just send the last rendered markup. It also would make sense that loading was true at some point.

**Comment by renato** [comment]
I'm facing the same issue when updating from 2.5.8 to 3.1.3. So if I need the initial server render to resolve all queries for SEO purposes, there's no workaround, the only solution currently is sticking to v2?

**Comment by NinjaOnRails** [comment]
> I'm facing the same issue when updating from 2.5.8 to 3.1.3. So if I need the initial server render to resolve all queries for SEO purposes, there's no workaround, the only solution currently is sticking to v2?

I use it with Next JS and the temporary fix is to call the queries and pass them to the component with getInitialProps. This is how I get the data for SSR

**Comment by flippidippi** [comment]
I was able to resolve the initial SSR issue, it now comes back as loaded, but we are still having an issue when changing routes and the "SSR" not working before changing routes.

Example repo: https://github.com/flippidippi/next-apollo-app

**Comment by bdew** [comment]
> To workaround this, React Apollo 3 currently returns a `loading: true` based result for the first SSR pass, which is accurately mirroring what's really happening behind the scenes. Unfortunately, as mentioned in several comments in this thread, it can lead to un-necessary `loading: true` states being shipped to the client.

I don't think this is an accurate description of the problem. Those `loading: true` results were working fine with hooks in apollo client 2, the problem isn't with the them, the problem is that they never actually change to the query result and the promise returned from getDataFromTree resolves with everything still being in the loading state.

The way it works with 2 is they return `loading: true` and then the query resolves and re-renders the component with the fetched data, and does so recursively till no more queries are loading. Only after that getDataFromTree promise resolves and data is sent to client.

I don't see why you need to introduce another null result state to retain that functionality.

**Comment by EyMaddis** [comment]
In my setup the backend seems to load everything just fine, meaning that the apollo cache state is always filled with the correct cache state. However, in reality, the serverside rendered result is showing the loading state. The client-side always renders the non-loading state because the cache is prefilled from the DOM.

I logged the loading states (on the server) and it is: `true, false, true` which is super weird. The first render is gathering the queries, the second one is the response, but the third is the Next.js rendering with a prefilled cache.
Right now I could only see this in development on my machine. Once it is that way every request is as such. I have to restart the server to fix it. I could not yet see this in production, but this is also just a staging environment.

I would love to share a reproducible repo but did not find a proper way to reliably reproduce this issue.
My current guess is that hot reload might mess stuff up after saving a couple of files.

I also debugged deep into apollo and could not see why loading is returned, because the cache was also valid for the SSR query handling (checked via the debugger).

**Comment by lifeiscontent** [comment]
@EyMaddis what is your `defaultFetchPolicy` / `fetchPolicy`? I found that `cache-and-network` did what you were saying, but `cache-first` worked as expected, with the exception that if you don't update the cached objects via mutations then you'll get stale data. In my own case, we had a `viewed matches` term which wasn't updating in the `cache-first` case due to it changing outside of Apollo (legacy code).

**Comment by carlosbaraza** [comment]
Coming from #3390 

I think the behavior of `@client` is quite confusing. I am trying to do something similar to what is [showcased in the docs](https://github.com/apollographql/apollo/blob/4ae53f9/docs/source/tutorial/local-state.md#query-local-data).

During SSR using Next and `getDataFromTree`:
1. First, it fetches the user and populates the cache using `writeData`.
2. When I run a `@client` query for the user data, it returns a `loading: true, data: {}`, instead of the user data it should return.

My workaround is reading the cache directly using `readQuery`, which seems to return the data without that `loading: true` step.

**Comment by EyMaddis** [comment]
@lifeiscontent I tried multiple strategies, but they do not provide any change. I also made sure (using cache-only) that the default settings are applied (because the app does not load anything then).

**Comment by lifeiscontent** [comment]
@EyMaddis Can you provide a minimal reproducible use case?

**Comment by EyMaddis** [comment]
@lifeiscontent I would love to, unfortunately this only happens seemingly at random while developing. My current guess is that the auto-reload while saving files (in a Next.js project) might break this after a lot of edits.
It might also be a memory issue (+ a leak) that can lead to this, but I tried to play around with max memory limits without any noticeable change.

**Comment by krzksz** [comment]
Hey, I've just stumbled across this bug and it seems like reusing local state variables inside queries using `@export(as: "variableName")`, here's a reproduction where I'm extracting `skip` variable from the cache:
https://codesandbox.io/s/with-apollo-r3599?fontsize=14&hidenavigation=1&theme=dark

**Comment by hwillson** [comment]
This issue thread has gone in a few different directions which makes it difficult to extract (and track) actionable items. I'll reply to some of the oustanding comments below (the ones that I think encompass the bulk of the questions), but to anyone still having an issue with this, please re-read my comments in https://github.com/apollographql/react-apollo/issues/3338#issuecomment-528462945 to understand why React Apollo v2 and v3 handle SSR differently.

@seantimm Regarding https://github.com/apollographql/react-apollo/issues/3338#issuecomment-528728563, I'd love to hear more. Would you be able to open this in a new issue, possibly with a reproduction?

@kedarguy Regarding https://github.com/apollographql/react-apollo/issues/3338#issuecomment-529196439, first off thanks for the reproduction. What this reproduction shows however is exactly what I've explained in https://github.com/apollographql/react-apollo/issues/3338#issuecomment-528462945. There is an additional loading state happening on the server side, that needs to be handled in your code. You can block the extra loading state from being rendered (which is ultimately leading to the destructuring error you're seeing. There are different ways to address this, but one quick way is to adjust your `WithApollo` component to return a `ssrComplete: true` prop in `getInitialProps` like:

```js
WithApollo.getInitialProps = async ctx => {
  ...
  return {
    ...pageProps,
    apolloState,
    ssrComplete: true
  }
}
```

then further adjust your `WithApollo` component to leverage this new prop to know when to render your component tree:

```js
const WithApollo = ({ apolloClient, apolloState, ssrComplete, ...pageProps }) => {
  const client = useMemo(
    () => apolloClient || initApolloClient(apolloState),
    []
  );

  if (!ssrComplete) {
    return null;
  }

  return (
    <ApolloProvider client={client}>
      <PageComponent {...pageProps} />
    </ApolloProvider>
  );
}
```

As mentioned in https://github.com/apollographql/react-apollo/issues/3338#issuecomment-528462945 I'm all ears if people would like a different solution implemented, but as things stand that initial loading state is required.

@bdew Regarding https://github.com/apollographql/react-apollo/issues/3338#issuecomment-560304562, the loading `true` results were working with AC 2 because of the reasons I outlined in https://github.com/apollographql/react-apollo/issues/3338#issuecomment-528462945 (in the `Background` section). The suggested workaround I mentioned was just an idea, and I'm open to other suggestions here. This isn't a bug though. The issue is caused by architecture differences between v2 and v3, due to the transition from using render props for everything, to hooks.

@krzksz Regarding https://github.com/apollographql/react-apollo/issues/3338#issuecomment-570085449, I updated your repro to use version 3.1.3 of `@apollo/react-hooks` and `@apollo/react-ssr`, and it appears to work properly. If it doesn't, please open this in a new issue.

Given that this issue thread has grown a bit unwiedly, I'll close it off. Please open new issues for further SSR related problems. Thanks!

**Comment by debugpai** [comment]
I was facing problem with this issue as well. But I managed to get the final server side rendered state to show the results by just using the rendered string from `getDataFromTree` instead of calling `renderToString` as mentioned in the docs.

```js
const html = await getDataFromTree(App)

```

instead of 

```js
await getDataFromTree(App);
const html = renderToString(App);
```
